File: frontend/lib/schemas/invoices.ts
import { z } from "zod";

export const invoiceSourceSchema = z.enum([
  "manual",
  "aade",
  "ocr_pdf",
  "csv_import",
  "gmail",
]);

export type InvoiceSource = z.infer<typeof invoiceSourceSchema>;

export const invoiceStatusSchema = z.enum([
  "draft",
  "pending_review",
  "finalized",
  "synced",
]);

export type InvoiceStatus = z.infer<typeof invoiceStatusSchema>;

export const invoiceAccountingKindSchema = z.enum([
  "ap_expense",
  "ar_revenue",
  "unknown",
]);

export type InvoiceAccountingKind = z.infer<typeof invoiceAccountingKindSchema>;

export const invoiceResponseSchema = z.object({
  id: z.string(),
  organization_id: z.string(),
  source: invoiceSourceSchema,
  external_id: z.string().nullable(),
  counterparty_id: z.string().nullable(),
  counterparty_display_name: z.string().nullable(),
  amount: z.coerce.number(),
  currency: z.string(),
  issue_date: z.string(),
  due_date: z.string().nullable(),
  status: invoiceStatusSchema,
  confidence: z.number().min(0).max(1).nullable(),
  requires_human_review: z.boolean(),
  is_recurring: z.boolean().optional().default(false),
  gl_journal_entry_id: z.string().uuid().nullish(),
  accounting_kind: invoiceAccountingKindSchema.nullish(),
});

export type InvoiceResponse = z.infer<typeof invoiceResponseSchema>;

export const invoiceCreateSchema = z.object({
  source: invoiceSourceSchema.default("manual"),
  external_id: z.string().max(255).nullable().optional(),
  counterparty_id: z.string().uuid().nullable().optional(),
  counterparty_display_name: z.string().max(255).nullable().optional(),
  customer_name: z.string().max(255).nullable().optional(),
  amount: z.coerce.number().positive(),
  currency: z.string().length(3).default("EUR"),
  issue_date: z.string().min(1),
  due_date: z.string().nullable().optional(),
  status: invoiceStatusSchema.default("draft"),
  confidence: z.number().min(0).max(1).nullable().optional(),
  requires_human_review: z.boolean().optional().default(false),
});

export type InvoiceCreate = z.infer<typeof invoiceCreateSchema>;

/** @deprecated Use invoiceCreateSchema / InvoiceCreate */
export const manualInvoiceCreateSchema = invoiceCreateSchema;
/** @deprecated Use invoiceResponseSchema / InvoiceResponse */
export const manualInvoiceResponseSchema = invoiceResponseSchema;
export type ManualInvoiceCreate = InvoiceCreate;
export type ManualInvoiceResponse = InvoiceResponse;

/** Form-only: amount entered as text then coerced for API. */
export const manualInvoiceSheetFormSchema = z.object({
  customer_name: z.string().min(1, "Customer name is required").max(255),
  amount: z
    .string()
    .min(1, "Amount is required")
    .refine((s) => {
      const n = Number.parseFloat(s.replace(",", "."));
      return Number.isFinite(n) && n > 0;
    }, "Enter a valid amount greater than zero"),
  issue_date: z.string().min(1, "Date is required"),
  currency: z.string().length(3).default("EUR"),
});

export type ManualInvoiceSheetFormValues = z.infer<
  typeof manualInvoiceSheetFormSchema
>;

export function invoiceCreateFromSheet(
  v: ManualInvoiceSheetFormValues
): InvoiceCreate {
  const amount = Number.parseFloat(v.amount.replace(",", "."));
  return invoiceCreateSchema.parse({
    source: "manual",
    counterparty_display_name: v.customer_name,
    amount,
    issue_date: v.issue_date,
    currency: v.currency,
  });
}

/** @deprecated Use invoiceCreateFromSheet */
export const manualInvoiceCreateFromSheet = invoiceCreateFromSheet;

File: backend/app/models/invoices.py
"""Pydantic request/response schemas for the unified ``invoices`` API."""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, model_validator


class InvoiceSourceEnum(str, Enum):
    """Wire format matches ``InvoiceSource`` ORM enum values."""

    MANUAL = "manual"
    AADE = "aade"
    OCR_PDF = "ocr_pdf"
    CSV_IMPORT = "csv_import"
    GMAIL = "gmail"


class InvoiceStatusEnum(str, Enum):
    """Wire format matches ``InvoiceStatus`` ORM enum values."""

    DRAFT = "draft"
    PENDING_REVIEW = "pending_review"
    FINALIZED = "finalized"
    SYNCED = "synced"


class InvoiceAccountingKindEnum(str, Enum):
    """Wire format matches ``InvoiceAccountingKind`` ORM enum values."""

    AP_EXPENSE = "ap_expense"
    AR_REVENUE = "ar_revenue"
    UNKNOWN = "unknown"


class InvoiceCreateRequest(BaseModel):
    """Create an invoice row (dashboard manual entry by default)."""

    source: InvoiceSourceEnum = InvoiceSourceEnum.MANUAL
    external_id: str | None = Field(None, max_length=255)
    counterparty_id: str | None = Field(None, description="UUID of linked counterparty")
    counterparty_display_name: str | None = Field(None, max_length=255)
    customer_name: str | None = Field(
        None,
        max_length=255,
        description="Alias for counterparty_display_name (frontend convenience)",
    )
    amount: Decimal = Field(..., gt=0)
    currency: str = Field(default="EUR", min_length=3, max_length=3)
    issue_date: date
    due_date: date | None = None
    status: InvoiceStatusEnum = InvoiceStatusEnum.DRAFT
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    requires_human_review: bool = False
    is_recurring: bool = False

    @model_validator(mode="after")
    def _manual_requires_label(self) -> InvoiceCreateRequest:
        if self.source == InvoiceSourceEnum.MANUAL and not self.counterparty_id:
            label = (
                (self.counterparty_display_name or "").strip()
                or (self.customer_name or "").strip()
            )
            if not label:
                raise ValueError(
                    "Manual invoices require counterparty_display_name or customer_name "
                    "when counterparty_id is omitted."
                )
        return self

    def resolved_counterparty_display_name(self) -> str | None:
        return (
            (self.counterparty_display_name or "").strip()
            or (self.customer_name or "").strip()
            or None
        )


class InvoiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    organization_id: str
    source: InvoiceSourceEnum
    external_id: str | None
    counterparty_id: str | None
    counterparty_display_name: str | None
    amount: Decimal
    currency: str
    issue_date: date
    due_date: date | None
    status: InvoiceStatusEnum
    confidence: float | None
    requires_human_review: bool
    is_recurring: bool
    gl_journal_entry_id: str | None = None
    accounting_kind: InvoiceAccountingKindEnum | None = None
